<html>
<head>
    <title>SPAdes 2.5.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.5.0 Manual</h1>

1. <a href="#sec1">About SPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.1. <a href="#sec1.1">Supported data types</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.2. <a href="#sec1.2">SPAdes pipeline</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.3. <a href="#sec1.3">SPAdes performance</a><br>
2. <a href="#sec2">Installation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#sec2.1">Downloading SPAdes Linux binaries</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.2. <a href="#sec2.2">Downloading SPAdes binaries for Mac</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.3. <a href="#sec2.3">Downloading and compiling SPAdes source code</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.4. <a href="#sec2.4">Verifying your installation</a><br>
3. <a href="#sec3">Running SPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#sec3.1">SPAdes input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#sec3.2">SPAdes command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#sec3.3">Assembling long Illumina paired reads (2x150 and 2x250)</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.4. <a href="#sec3.4">SPAdes output</a><br>

&nbsp;&nbsp;&nbsp;&nbsp;3.5. <a href="#sec3.5">Assembly evaluation</a><br>
4. <a href="#sec4">Citation</a><br>
5. <a href="#sec5">Feedback and bug reports</a><br>
<br>

<a name="sec1"></a>
<h2>1. About SPAdes</h2>
<p>
    SPAdes &ndash; St. Petersburg genome assembler &ndash; is intended for both standard isolates and single-cell MDA bacteria assemblies. This manual will help you to install and run SPAdes. SPAdes version 2.5.0 was released under GPLv2 on July 4, 2013 and can be downloaded from  <a href="http://bioinf.spbau.ru/spades" target="_blank">http://bioinf.spbau.ru/spades</a>.


<a name="sec1.1"></a>
<h3>1.1 Supported data types</h3>
<p>
    The current version of SPAdes works only with Illumina reads. Support for other technologies (e.g.  Roche 454, IonTorrent, PacBio) is currently in progress and will probably be included in one of the next releases. 
<p>
    Version 2.5.0 of SPAdes supports paired-end reads, mate-pairs and unpaired reads. SPAdes can take as input several paired-end and mate-pair libraries simultaneously. 
<p>
    Also note, that SPAdes was initially designed for single-cell and standard bacterial data sets and is not intended for larger genomes (e.g. mammalian size genomes) and metagenomic projects. For such purposes you can use it at your own risk.


<a name="sec1.2"></a>
<h3>1.2 SPAdes pipeline</h3>
<p>
SPAdes comes in three separate modules:
    <ul>
        <li> <a href="http://bioinf.spbau.ru/en/spades/bayeshammer" target="_blank">BayesHammer</a> &ndash; read error correction tool, which works well on both single-cell and standard data sets </li>
        <li> SPAdes &ndash; iterative short-read genome assembly module; by default consecutively iterates through K value equal to 21, 33 and 55 </li>
        <li> MismatchCorrector &ndash; a tool which improves mismatch and short indel rates in resulting contigs and scaffolds; this module uses the <a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> tool [<a href="http://www.ncbi.nlm.nih.gov/pubmed/19451168" target="_blank">Li H. and Durbin R., 2009</a>]; MismatchCorrector is turned off by default, but we recommend to turn it on (see <a href="#correctoropt">SPAdes options section</a>)</li>
    </ul>
<p>
    We recommend to run SPAdes with BayesHammer to obtain high-quality assemblies. However, if you use your own read correction tool it is possible to turn BayesHammer off. It is also possible to use only the read error correction stage, if you wish to use another assembler. See the <a href="#pipelineopt">SPAdes options section</a>.


<a name="sec1.3"></a>
<h3>1.3 SPAdes' performance</h3>
<p>
    In this section we give approximate data about SPAdes' performance on two data sets:
    <ul>
        <li> <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_mc/" target="_blank">Standard isolate <i>E. coli</i></a>; 6.2Gb, 28M reads, 2x100bp, insert size ~ 215bp </li>
        <li> <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_sc/" target="_blank">MDA single-cell <i>E. coli</i></a>; 6.3 Gb, 29M reads, 2x100bp, insert size ~ 270bp </li>
    </ul>
<p>
    We ran SPAdes with default parameters using 16 threads on a server with Intel Xeon 2.27GHz processors. BayesHammer runs in approximately 1 hour and takes up to 15Gb of RAM to perform read error correction on each data set. Assembly takes 35 minutes and 4Gb of RAM for the <i>E. coli</i> isolate data set. The <i>E. coli</i> single-cell data set takes 55 minutes and 6Gb of RAM. MismatchCorrector runs for about an hour on standard <i>E. coli</i>, about 2 hours on the single-cell data set, and requires 8Gb of RAM. All modules also require additional disk space for temporary files. See the table below for more precise values.

<p>
    <table border="1" cellpadding="4" cellspacing="0">
        <tr>
            <td align="right"> Data set &nbsp; </td>
            <td colspan="3" align="center"> <i>E.coli</i> isolate </td> 
            <td colspan="3" align="center"> <i>E.coli</i> single-cell </td>
        </tr>

        <tr>
            <td> Stage </td>
            <td align="center"> &nbsp; Time (min) &nbsp; </td> 
            <td align="center"> &nbsp; Peak RAM &nbsp;<br>&nbsp; usage (Gb) &nbsp; </td>
            <td align="center"> &nbsp; Additional &nbsp;<br>&nbsp; disk space (Gb) &nbsp; </td>
            <td align="center"> &nbsp; Time (min) &nbsp; </td> 
            <td align="center"> &nbsp; Peak RAM &nbsp;<br>&nbsp; usage (Gb) &nbsp; </td>
            <td align="center"> &nbsp; Additional &nbsp;<br>&nbsp; disk space (Gb) &nbsp; </td>
        </tr>

        <tr>
            <td> BayesHammer </td>
            <td align="center"> 75 </td> 
            <td align="center"> 14 </td>
            <td align="center"> 13 </td>
            <td align="center"> 75 </td> 
            <td align="center"> 15 </td>
            <td align="center"> 13 </td>
        </tr>

        <tr>
            <td> SPAdes </td>
            <td align="center"> 35 </td> 
            <td align="center"> 4 </td>
            <td align="center"> 8 </td>
            <td align="center"> 55 </td> 
            <td align="center"> 6 </td>
            <td align="center"> 8 </td>
        </tr>

        <tr>
            <td> MismatchCorrector </td>
            <td align="center"> 60 </td> 
            <td align="center"> 8 </td>
            <td align="center"> 24 </td>
            <td align="center"> 120 </td> 
            <td align="center"> 8 </td>
            <td align="center"> 24 </td>
        </tr>
    </table>

<p>
    Notes:
    <ul>
        <li> Running SPAdes without preliminary read error correction (e.g. without BayesHammer) will likely require more time and memory </li>
        <li> Each module removes its temporary files as soon as it finishes </li>
    </ul>


<a name="sec2"></a>
<h2>2. Installation</h2>
<p>
    SPAdes requires a 64-bit Linux system or Mac OS and Python 2.4 or higher (but not 3) to be pre-installed on it. To obtain SPAdes you can either download binaries or download source code and compile it yourself.

<a name="sec2.1"></a>
<h3>2.1 Downloading SPAdes Linux binaries</h3>

<p>
    To download <a href="http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0-Linux.tar.gz">SPAdes Linux binaries</a> and extract them, run:

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0-Linux.tar.gz
    tar -xzf SPAdes-2.5.0-Linux.tar.gz
    cd SPAdes-2.5.0-Linux/bin/
</code>
</pre>

<p>
    In this case you do not need to run any installation scripts &ndash; SPAdes is ready to use. The following files will be placed in <code>bin</code> directory:
    <ul>
        <li><code>spades.py</code>  (main executable script)</li>
        <li><code>hammer</code>  (read error correcting module)</li>
        <li><code>spades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>
 
<a name="sec2.2"></a>
<h3>2.2 Downloading SPAdes binaries for Mac</h3>

<p>
    To obtain <a href="http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0-Darwin.tar.gz">SPAdes binaries for Mac</a>, run:

<pre  class="code">
<code>
    curl http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0-Darwin.tar.gz -o SPAdes-2.5.0-Darwin.tar.gz
    tar -zxvf SPAdes-2.5.0-Darwin.tar.gz
    cd SPAdes-2.5.0-Darwin/bin/
</code>
</pre>

<p>
    Just as in Linux, SPAdes is ready to use and no further installation steps are required. You will get the same files in <code>bin</code> directory:
    <ul>
        <li><code>spades.py</code>  (main executable script)</li>
        <li><code>hammer</code>  (read error correcting module)</li>
        <li><code>spades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>
 
<a name="sec2.3"></a>
<h3>2.3 Downloading and compiling SPAdes source code</h3>
<p>
    If you wish to compile SPAdes by yourself you will need the following libraries to be pre-installed:
    <ul>
        <li>g++ (version 4.4 or higher)</li>
        <li>cmake (version 2.8 or higher)</li>
        <li>zlib</li>
    </ul>

<p>
    If you meet these requirements, you can download <a href="http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0.tar.gz">SPAdes source code</a>: 

<pre class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.5.0/SPAdes-2.5.0.tar.gz
    tar -xzf SPAdes-2.5.0.tar.gz
    cd SPAdes-2.5.0
</code>
</pre>

<p>
    and build it with the following script:

<pre class="code">
<code>
    ./spades_compile.sh
</code>
</pre>

<p> 
    In this case SPAdes will be built into <code>./bin</code> directory. If you wish to install SPAdes into another directory, you can specify full path of destination folder by running:

<pre class="code">
<code>
    PREFIX=&lt;destination_dir> ./spades_compile.sh
</code>
</pre>

<p>
    for example:

<pre class="code">
<code>
    PREFIX=/usr/local ./spades_compile.sh
</code>
</pre>

<p>
   which will install SPAdes into <code>/usr/local/bin</code>.
    
<p>
    After installation you will get the same files in <code>./bin</code> (or <code>&lt;destination_dir>/bin</code> if you specified PREFIX) directory:
    <ul>
        <li><code>spades.py</code>  (main executable script)</li>
        <li><code>hammer</code>  (read error correcting module)</li>
        <li><code>spades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>

<a name="sec2.4"></a>
<h3>2.3 Verifying your installation</h3>
<p>
    For testing purposes, SPAdes comes with a &ldquo;toy data set&rdquo; (reads, that align to first 1000 bp of <i>E. coli</i>). To try SPAdes on this data set run:

<pre class="code">
<code>
    ./spades.py --test
</code>
</pre>

<p>
    If the installation is successful, you will find the following information at the end of the log:

<pre class="code">
<code>
 * Corrected reads are in spades_test/corrected/
 * Assembled contigs are in spades_test/contigs.fasta
 * Assembled scaffolds are in spades_test/scaffolds.fasta

Thank you for using SPAdes!

======= SPAdes pipeline finished. Log can be found here: spades_test/spades.log
</code>
</pre>



<a name="sec3"></a>
<h2>3. Running SPAdes</h2>

<a name="sec3.1"></a>
<h3>3.1 SPAdes input</h3>
<p>
    SPAdes takes as input paired-end reads, mate-pairs and single (unpaired) reads in FASTA or FASTQ format. However, in order to run read error correction, reads should be in FASTQ format.

<p>
    By using command line interface you can specify up to five different paired-end libraries and also up to five different mate-pair ones (if you wish to use more, please contact us <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>). The number of unpaired libraries is unlimited. SPAdes 2.5 should not be used if only mate-pair libraries are available. We further refer to paired-end and mate-pair libraries simply as to read-pair libraries.

<p>
    By default SPAdes assumes that paired-end reads have forward-reverse (fr) orientation and mate-pairs have reverse-forward (rf) one. However, different orientations can be set for any library by using SPAdes options. 

<p>
    To distinguish reads in pairs we refer to them as left and right reads. For forward-reverse orientation forward reads correspond to the left reads and reverse reads &ndash; to the right. Similarly, in reverse-forward orientation left and right reads correspond to reverse and forward reads respectively etc. 

<p>
    Each read-pair library can be stored in several files or several pairs of files. Paired reads can be organized in two different ways:

    <ul>
        <li> In file pairs. In this case left and right reads are placed in different files and go in the same order in respective files. </li>
        <li> In merged files. In this case, the reads are interlaced, so that each right read goes after the corresponding paired left read. </li>
    </ul>

<p>
    For example, Illumina produces paired-end reads in two files: <code>s_1_1_sequence.txt</code> and <code>s_1_2_sequence.txt</code>. If you choose to store reads in file pairs make sure that for every read from <code>s_1_1_sequence.txt</code> the corresponding paired read from <code>s_1_2_sequence.txt</code> is placed in the respective paired file on the same line number. If you choose to use merged files, every read from <code>s_1_1_sequence.txt</code> should be followed by the corresponding paired read from <code>s_1_2_sequence.txt</code>.

<p>
    Note that SPAdes also accepts gzip-compressed files.

<a name="sec3.2"></a>
<h3>3.2 SPAdes command line options </h3>
<p>
    To run SPAdes from the command line, type 

<pre class="code">
<code>
    ./spades.py [options] -o &lt;output_dir>
</code>
</pre>

<a name="basicopt"></a>
<h4>Basic options</h4>
<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Required option.
</p>

<p>
    <code>--sc </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;This flag is required for MDA (single-cell) data.
</p>

<p>
    <code>--test</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the &ldquo;toy data set&rdquo;; see <a href="#sec2.3">section 2.3</a>.
</p>

<p>
    <code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.
</p>


<a name="inputdata"></a>
<h4>Input data</h4>

<i><b>&nbsp;&nbsp;Specifying single library (previously used format)</i></b>
<p>
    <code>--12 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with interlaced forward and reverse paired-end reads.
</p>

<p>
    <code>-1 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with forward reads.
</p>

<p>
    <code>-2 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with reverse reads.
</p>

<p>
    <code>-s &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with unpaired reads.
</p>

<i><b>&nbsp;&nbsp;Specifying multiple libraries (new format)</i></b>
<p>
    <code>--pe<b>&lt;#></b>-12 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with interlaced reads for paired-end library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5). For example, for the first paired-end library the option is:
    <code>--pe1-12 &lt;file_name> </code><br>
</p>

<p>
    <code>--pe<b>&lt;#></b>-1 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with left reads for paired-end library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5).
</p>

<p>
    <code>--pe<b>&lt;#></b>-2 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with right reads for paired-end library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5).
</p>

<p>
    <code>--pe<b>&lt;#></b>-s &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with unpaired reads from paired-end library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;For exmaple, paired reads can become unpaired during error correction procedure.
</p>

<p>
    <code>--pe<b>&lt;#></b>-<b>&lt;or></b> &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Orientation of reads for paired-end library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5; <code><b>&lt;or></b></code> = "fr","rf","ff"). <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Default orientation for paired-end libraries is forward-reverse. For example, to specify reverse-forward orientation for the second paired-end library you should use flag:
    <code>--pe2-rf </code><br>
</p>
<p>
    <code>--mp<b>&lt;#></b>-12 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with interlaced reads for mate-pair library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5).
</p>

<p>
    <code>--mp<b>&lt;#></b>-1 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with leftreads for mate-pair library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5).
</p>

<p>
    <code>--mp<b>&lt;#></b>-2 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with right reads for mate-pair library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5).
</p>
<p>
    <code>--mp<b>&lt;#></b>-<b>&lt;or></b> &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Orientation of reads for mate-pair library number <code><b>&lt;#></b></code> (<code><b>&lt;#></b></code> = 1,2,3,4,5; <code><b>&lt;or></b></code> = "fr","rf","ff"). <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Default orientation for mate-pair libraries is reverse-forward. For example, to specify forward-forward orientation for the first mate-pair library you should use flag:
    <code>--mp1-ff </code><br>
</p>

<a name="pipelineopt">
<h4>Pipeline options</h4>

<p>
    <code>--only-error-correction</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Perfrorms read error correction only.
</p>

<p>
    <code>--only-assembler</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs assembly module only.
</p>

<a name="correctoropt">
<p>
    <code>--careful</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Tries to reduce number of mismatches and short indels. Also runs MismatchCorrector &ndash; a post processing tool, which uses <a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> tool (comes with SPAdes). This option is recommended. 

</p>

<p>
    <code>--disable-gzip-output</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Forces read error correction module not to compress the corrected reads. If this options is not set, corrected reads will be in <code>*.fastq.gz</code> format.
</p>

<p>
    <code>--rectangles</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Uses <a href="http://bioinf.spbau.ru/en/rectangles" target="_blank">rectangle graph algorithm</a> for repeat resolution stage instead of usual SPAdes repeat resolution module (experimental).
</p>

<a name="advancedopt">
<h4>Advanced options</h4>

<p>
    <code>-t &lt;int></code> (or <code>--threads &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of threads. The default value is 16.
</p>

<p>
    <code>-m &lt;int></code> (or <code>--memory &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set memory limit in Gb. SPAdes terminates if it reaches this limit. The default value is 250 Gb. Actual amount of consumed RAM will be below this limit.
</p>

<p>
    <code>--tmp-dir &lt;dir_name></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set directory for temporary files from read error correction. The default value is <code>&lt;output_dir>/corrected/tmp</code>.
</p>

<p>
    <code>-k &lt;int,int,...></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Comma-separated list of k-mer sizes to be used (all values must be odd, less than 128 and listed in ascending order). The default value is 21,33,55.
</p>

<p>
    <code>-i &lt;int></code> (or <code>--iterations &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of iterations for read error correction. The default value is 1.
</p>

<p>
    <code>--phred-offset &lt;33 or 64></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;PHRED quality offset for the input reads, can be either 33 or 64. It will be auto-detected if it is not specified.
</p>

<p>
    <code>--debug</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes in debug mode, keeping intermediate results.
</p>

<a name="examples">
<h4>Examples</h4>
<p>
    To test &ldquo;toy data set&rdquo; you can also run:

<pre class="code">
<code>
    ./spades.py --pe1-1 ../share/spades/test_dataset/ecoli_1K_1.fq.gz --pe1-2 ../share/spades/test_dataset/ecoli_1K_2.fq.gz -o spades_test
</code>
</pre>

<p>
    If you have your library separated into several pairs of files, for example:

<pre class="code">
<code>
    lib1_forward_1.fastq
    lib1_reverse_1.fastq
    lib1_forward_2.fastq
    lib1_reverse_2.fastq
</code>
</pre>


<p>
    make sure that corresponding files are given in the same order:

<pre class="code">
<code>
    ./spades.py --pe1-1 lib1_forward_1.fastq --pe1-2 lib1_reverse_1.fastq --pe1-1 lib1_forward_2.fastq --pe1-2 lib1_reverse_2.fastq -o spades_output
</code>
</pre>

<p>
    Files with interlacing paired-end reads or files with unpaired reads can be specified in any order with one file per option, for example:

<pre class="code">
<code>
    ./spades.py --pe1-12 lib1_1.fastq --pe1-12 lib1_2.fastq --pe1-s lib1_unpaired_1.fastq --pe1-s lib1_unpaired_2.fastq -o spades_output    
</code>
</pre>

<p>
    In case you have several paired-end and mate-pair reads, for example:

<li> paired-end library 1
<pre class="code">
<code>
    lib_pe1_left.fastq
    lib_pe1_right.fastq
</code>
</pre>
<p>
<li> mate-pair library 1

<pre class="code">
<code>
    lib_mp1_left.fastq
    lib_mp1_right.fastq
</code>
</pre>
<p>
<li> mate-pair library 2
<pre class="code">
<code>
    lib_mp2_left.fastq
    lib_mp2_right.fastq
</code>
</pre>

<p>
    make sure that files corresponding to each library are grouped together:

<pre class="code">
<code>
    ./spades.py --pe1-1 lib_pe1_left.fastq --pe1-2 lib_pe1_right.fastq --mp1-1 lib_mp1_left.fastq --mp1-2 lib_mp1_right.fastq --mp2-1 lib_mp2_left.fastq --mp2-2 lib_mp2_right.fastq -o spades_output
</code>
</pre>

<p>
    All options for specifying input data can be mixed if needed, but make sure that files for each library are grouped and files with left and right paired reads are listed in the same order.


<a name="sec3.3">
<h3>3.3 Assembling long Illumina paired reads (2x150 and 2x250)</h3>
<p>
Recent advances in DNA sequencing technology led to rapid increase of a read length. Nowadays it is a common situation to have a dataset consisting of 2x150 or 2x250 paired-end reads produced by Illumina MiSeq or HiSeq2500. However, the use of longer reads alone will not automatically improve assembly quality. Proper assembler that can make use of all their advantages is needed.
<p>
As far as SPAdes uses iterative k-mer length, it allows to benefit from the full potential of the long paired-end reads. Currently one has to set the assembler options up manually but we plan to incorporate automatic calculation of necessary options soon.
<p>
Please note that not only the read length matters, but insert length does matter a lot. It is suboptimal to sequence 300bp fragment into a pair of 250bp reads. We suggest using 350-500 bp fragments with 2x150 reads and 550-700 bp fragments with 2x250 reads.

<h4>Multi-cell dataset with read length 2 x 150</h4>
<p>
Make sure your reads are corrected prior to assembly with Quake (recommended), or BayesHammer (integrated into SPAdes pipeline).
<p>
The default selection of k-mer lengths is 21, 33, 55 and might work well. If you have enough coverage (50x+), then you may want to try to set k-mer lengths of 21, 33, 55, 77.
<p>
Make sure you run assembler with <code>--careful</code> option to minimize number of mismatches in the final contigs.
<p>
We recommend you to check the SPAdes log file at the end of the each iteration to control the average coverage of the contigs.

<p>
For reads corrected prior to assembly run: 
<pre class="code">
<code>
    ./spades.py -k 21,33,55,77 --careful --only-assembler &lt;your reads> -o spades_output
</code>
</pre>
 
<p>
For non-corrected reads run:
<pre class="code">
<code>
    ./spades.py -k 21,33,55,77 --careful r &lt;your reads> -o spades_output
</code>
</pre>

<h4>Multi-cell dataset with read lengths 2 x 250</h4>
<p>
Make sure your reads are corrected prior to assembly with Quake (recommended), or BayesHammer (integrated into SPAdes pipeline).
<p>
By default we suggest to increase k-mer lengths in increments of 22 until the  k-mer length reaches 127. The exact length of the k-mer depends on the coverage: k-mer length of 127 corresponds to 50x k-mer coverage and higher.
<p>
Make sure you run assembler with <code>--careful</code> option to minimize number of mismatches in the final contigs.
<p>
We recommend you to check the SPAdes log file at the end of the each iteration to control the average coverage of the contigs.
<p>
For reads corrected prior to assembly run: 
<pre class="code">
<code>
    ./spades.py -k 21,33,55,77,99,127 --careful --only-assembler &lt;your reads> -o spades_output
</code>
</pre>
 
<p>
For non-corrected reads run:
<pre class="code">
<code>
    ./spades.py -k 21,33,55,77,99,127 --careful r &lt;your reads> -o spades_output
</code>
</pre>

<h4>Single-cell dataset with read lengths 2 x 150 or 2 x 250</h4>
<p>
The default options are recommended.
<p>
However, it might be tricky to fully utilize the advantages of  long reads you have. Consider contacting us for more information & discussions of assembling strategy.
<br>


<a name="sec3.4">
<h3>3.4 SPAdes output</h3>
<p>
    SPAdes stores all output files in <code>&lt;output_dir> </code>, which is set by user.

    <ul>
        <li><code>&lt;output_dir>/corrected/</code> directory contains reads corrected by BayesHammer in <code>*.fastq.gz</code> files; if compression is disabled reads are stored in usual <code>*.fastq</code> files</li>
        <li><code>&lt;output_dir>/contigs.fasta</code> contains resulting contigs</li>
        <li><code>&lt;output_dir>/scaffolds.fasta</code> contains resulting scaffolds</li>
    </ul>

<p>

<p>
    Notes:
    <ul>
        <li> Scaffolds are not produced if no read-pair library is provided </li>
        <li> Scaffolding procedure in SPAdes is optimized for paired-end reads, but not yet for mate-pair libraries </li>
        <li> You can use third-party scaffolder (e.g. <a href="http://sourceforge.net/projects/operasf/files/version%201.3.1/"  target="_blank">Opera</a>) if you have mate-pair libraries </li>
    </ul>

<p>
Full list of <code>&lt;output_dir></code> content is presented below: 
</p>
<pre>
    <code>contigs.fasta</code> &ndash; <i>resulting contigs</i>
    <code>scaffolds.fasta</code> &ndash; <i>resulting scaffolds (will not be produced if input reads are unpaired)</i>

    <code>corrected/</code> &ndash; <i>files from read error correction</i>
        <code>configs/</code> &ndash; <i>configuration files for read error correction</i>
        <code>dataset.info</code> &ndash; <i>internal configuration file</i>
        Output files with corrected reads

    <code>params.txt</code> &ndash; <i>information about SPAdes parameters in this run</i>
    <code>spades.log</code> &ndash; <i>SPAdes log</i>
    <code>dataset.info</code> &ndash; <i>internal configuration file</i>
    <code>K&lt;##>/</code> &ndash; <i>directory containing files from the run with K=&lt;##> (K21, K33 and K55 are created by default)</i>
</pre>

<p>
    SPAdes will overwrite these files and directories if they exist in the specified <code>&lt;output_dir></code>. 


<a name="sec3.5">
<h3>3.5 Assembly evaluation</h3>

<p>
    We recommend to use <a href="http://bioinf.spbau.ru/en/quast" target="_blank">QUAST</a> for assembly evaluation. 
<br>



<a name="sec4">
<h2>4. Citation</h2>
<p>
    If you use SPAdes in your research, please include <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021" target="_blank">Bankevich, Nurk et al., 2012</a> in your reference list.

<p>
    In addition, we would like to list your publications that use our software on our website. Please email the reference, the name of your lab, department and institution to <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>.
<br>


<a name="sec5">
<h2>5. Feedback and bug reports</h2>
<p>
    Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve SPAdes.

<p>
    If you have trouble running SPAdes, please provide us with the files <code>params.txt</code> and <code>spades.log</code> from the directory <code>&lt;output_dir></code>.

<p>
    Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>.

<br/><br/><br/><br/><br/>

</body>
</html>
